### PR TITLE
Includes test for dummy gradient and fix propagated bugs

### PR DESCRIPTION
**Pull request by facusapienza21**
** Do not merge yet **

@albangossard take a look to this PR and let me know if you have comments/suggestions. 

You will see the test I added currently don't pass due to some changes I did in Sleipnir that I would prefer to keep local for now. Tomorrow I will see if I can remove these (I need to convert \theta to a ComponentVector than a Vector, as it is right now). This won't take much time to change, but I am leaving the office soon and I did wanted to have second eyes on this as early as possible. I will make some more changes tomorrow and hopefully we can move forward from there.

### CODE REVIEW

**Comment by facusapienza21** [comment]
Unclear why some changes are relevant to the documentation since I thought I was working with the latest version of ODINN... please ignore these.

**Review by albangossard — COMMENTED** [review]
I didn't see changes that could break AD. The only thing I'm concerned about it type instability through the use of `Union`. I have been investigating that yesterday with `@code_warntype` and it turns out we have a lot of them. We should avoid using them as often as possible.

Please let me know when you want me to review this PR again

File: src/simulations/functional_inversions/callback_utils.jl [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+Callback function to track and diagose training. It includes print and updates in simulation::Simulation.
+"""
+function callback_diagnosis(θ, l, simulation)
+
+    # See if we want to change this or leave it
+    # update_training_state!(simulation, l)
+
+    push!(simulation.stats.losses, l)
+    if length(simulation.stats.losses) % 1 == 0
```
[the remark this answers is not in the document]
Comment (facusapienza21) [reply src/simulations/functional_inversions/callback_utils.jl:57]: That sounds like a good idea. Would that be like specifiying the verbose level between 0 and 4 and then everyprint just happens if their verbose leves is above the one prescribed? That sounds like something super simple to implement, happy to give it a try!

File: src/simulations/training_stats/TrainingStats.jl [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+mutable struct TrainingStats{F <: AbstractFloat, I <: Int}
+    retcode::Union{String, Nothing}
+    losses::Union{Vector{F}, Nothing}
+    niter::Union{I, Nothing}
+end 
+
+function TrainingStats(;
+    retcode::Union{String, Nothing} = nothing,
+    losses::Union{Vector{F}, Nothing} = Float64[],
+    niter::Union{I, Nothing} = 0
```
[the remark this answers is not in the document]
Comment (facusapienza21) [reply src/simulations/training_stats/TrainingStats.jl:12]: There is none, I just remove it!

File: src/simulations/training_stats/TrainingStats.jl [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+mutable struct TrainingStats{F <: AbstractFloat, I <: Int}
+    retcode::Union{String, Nothing}
+    losses::Union{Vector{F}, Nothing}
+    niter::Union{I, Nothing}
+end 
+
+function TrainingStats(;
+    retcode::Union{String, Nothing} = nothing,
+    losses::Union{Vector{F}, Nothing} = Float64[],
```
[the remark this answers is not in the document]
Comment (facusapienza21) [reply src/simulations/training_stats/TrainingStats.jl:11]: Also fix it!

File: src/simulations/functional_inversions/functional_inversion_utils.jl [hunk 4]
```diff
@@ -26,29 +29,50 @@ function train_UDE!(simulation::FunctionalInversion)
...
-    optprob = OptimizationProblem(optf, θ)
-    
-    if simulation.parameters.UDE.target == "A"
-        cb_plots(θ, l) = callback_plots_A(θ, l, simulation) # TODO: make this more customizable 
+    # Simplify API for optimization problem
+    loss_function(θ, p) = loss_iceflow(θ, p[1], p[2])
+
+    if isnothing(simulation.parameters.UDE.grad)
+        # optf = OptimizationFunction((θ, _, batch_ids, rgi_ids)->loss_iceflow(θ, batch_ids, simulation), simulation.parameters.UDE.optim_autoAD)
+        optf = OptimizationFunction((θ, p, _, _) -> loss_function(θ, p), simulation.parameters.UDE.optim_autoAD)
```
[the remark this answers is not in the document]
Comment (facusapienza21) [reply src/simulations/functional_inversions/functional_inversion_utils.jl:42]: Thank you so much @albangossard for asking this... I strugled a lot to make sense of it tomorrow, and after you comment I comeback to this point and find 1) an explanation and 2) a fix for this. I am adding a response for this below in the main conversation.

**Comment by facusapienza21** [comment]
@albangossard the reason for the multiple arguments in the loss function was that that was the only way of made it work with the code as it was. In principle, the API for optimization is something like 
```julia 
using Pkg; Pkg.activate(".")


import OptimizationOptimisers
using Optimization, Zygote
using SciMLBase: NoAD

rosenbrock(x, p) = (p[1] - x[1])^2 + p[2] * (x[2] - x[1]^2)^2 
x0 = zeros(2)
_p = (1.0, 100.0)

# rosenbrock(x) = (p[1] - x[1])^2 + p[2] * (x[2] - x[1]^2)^2


function rosenbrock_grad(dx, x, p)
    # dx = rand(Float64, size(x))
    dx[1] = 2 * (x[1] - p[1]) + 4 * p[2] * x[1] * (x[1]^2 - x[2])
    dx[2] = 2 * p[2] * (x[2] - x[1]^2)
end

optf = OptimizationFunction(rosenbrock, grad=rosenbrock_grad, NoAD())
# optf = OptimizationFunction(rosenbrock, AutoZygote())

prob = OptimizationProblem(optf, x0, _p)
sol = solve(prob, OptimizationOptimisers.Adam(0.05), maxiters = 1000, progress = false)
```
where the extra parameters `p` is use for hyperparameters. Now, in the code we are using minibatching via the `train_loaded` object. This requires that now `p` is the iterator of the batches of data. With my changes, I was putting all hyperparameters inside `p` and this makes the minibatching to work very weird (the two `_` arguments from the previous version where consequence of these extra arguments in the minibatching). After realizing this, I cleaned the API and now this works perfectly well with minibatching. Definitively it looks far more clean and correct now, so thank you for the comment!

**Comment by facusapienza21** [comment]
Unless you have further comments, feel free to merge this into `ODINN-SciML:fix-adjoints`, maybe we can make changes there and see if we can make the AD work here.

**Comment by facusapienza21** [comment]
Feel free to merge @albangossard ! The new test that I added are passing right now. Let me know if you have any more comments and I will address them.

File: src/simulations/functional_inversions/functional_inversion_utils.jl [hunk 5]
```diff
@@ -25,22 +28,45 @@ Trains UDE based on the current FunctionalInversion.
...
+    # Training diagnosis callback
+    cb_diagnosis(θ, l) = callback_diagnosis(θ, l, simulation)
+    # Combined callback
+    cb(θ, l) = CallbackOptimizationSet(θ, l; callbacks=(cb_plots, cb_diagnosis))
   
     println("Training iceflow UDE...")
     
     iceflow_trained = solve(optprob, 
                             simulation.parameters.hyper.optimizer, 
-                            ncycle(train_batches, simulation.parameters.hyper.epochs), allow_f_increases=true,
```
Comment (albangossard) [inline src/simulations/functional_inversions/functional_inversion_utils.jl:42]: In the context of stochastic optimization, isn't it problematic to remove `allow_f_increases=true`? The sequence of cost function values might have a lot of variability related to the sampling of the batches, and we don't want this variability to impact the returned solution.

**Review by albangossard — APPROVED** [review]
Thank you very much for this explanation @facusapienza21! Along with your changes, your explanation makes sense to me. From what I understand, we were concatenating the optimization variable with the parameters in a strange way, this is much cleaner now!
I left one last minor comment, but this is something we can fix afterwards if needed. I'll merge your PR so that I can start from it for the fix of the type instabilities.
